Complete Adaptive Prompting system

StudHamza opened the pull request:
# PR: Schema Modification accommodating the new Prompting system

## Schema modification Summary
Refactored the schema to transition from a static database to an **adaptive prompting engine**. Added `beebetter\test\database_test.dart` unit test.

<img width="1275" height="652" alt="Untitled" src="https://github.com/user-attachments/assets/5a5e9be9-ac40-44ef-ba0e-4fb33fa543fe" />


## Schema Changes

### 1. Enhanced Personalization (User Table)
* **Added:** `currentDifficultyLevel`, `journalingStreak`, `preferredCategories`, `avoidedPrompts`.
* **Rationale:** Enables gamification (streaks) and allows the system to scale prompt complexity based on user maturity.

### 2. Clinical & Contextual Metadata (Prompts Table)
* **Added:** `therapeuticFramework`, `difficultyLevel`, `targetMoodStates`, `bestTimeOfDay`, `tags`.
* **Rationale:** Supports evidence-based content (CBT/DBT) and allows for context-aware suggestions (e.g., specific prompts for morning vs. night).
* **Lifecycle:** Added `isActive` to allow soft-deletion of prompts.

### 3. Interaction Tracking (PromptInteractions Table - NEW)
* **Function:** Tracks `completed` vs `skipped` prompts.
* **Rationale:** Creates a feedback loop to identify which content resonates with users and refine the recommendation algorithm.

### 4. Simplified Taxonomy
* **Removed:** `Categories` and `PromptCategoryLinks` tables.
* **Rationale:** Flattened categorization into the `Prompts` table to reduce query complexity and improve performance.

---

| Feature | Old Version | New Version |
| :--- | :--- | :--- |
| **User Context** | Basic Profile | Behavioral & Preference Tracking |
| **Prompt Depth** | Content Only | Clinical Framework & Intent |
| **Feedback Loop** | None | Explicit Skip/Complete Tracking |
| **Complexity** | Normalized Categories | Flat, Metadata-rich Structure |

StudHamza committed:
Modified Schema

XHCFS commented on lib/data/database/tables.dart:
```diff
@@ -77,30 +78,44 @@ enum InputType {
+  TextColumn get bestTimeOfDay => text().nullable()();
+
+  // tags for filtering & clustering
+  TextColumn get tags => text().nullable()();
```
why not use categories and therapeutic framework

StudHamza committed:
Prompt Selection Scoring Algorithm

StudHamza committed:
Adaption system, with test cases

StudHamza committed:
Added 50 unique prompt based on different theruraptic therapeutic framework(3/3)

StudHamza commented:
# Issue Summary: Prompting System & Asset Implementation
This ticket closes #10 by implementing the full guided mode logic and the necessary database initialization services. Detailed instructions on how to use and implement these specific systems can be found in the README.md files located within each respective directory.

## 1. Prompting System Services (`/lib/prompting_system/services`)
This directory manages the core logic for the guided journaling mode.

### `prompt_initializer.dart` (New)
Handles the population of the database from local assets.
* **Functionality**: Parses `prompts.json`, maps string moods to `Mood` enum indexes, and performs batch inserts.
* **Unit Tests**: Verified in `test/prompt_initialization_test.dart` covering:
    * **Enum Mapping**: Converts string names (e.g., "proud") to integer indexes.
    * **Idempotency**: Prevents redundant writes using `isNotEmpty` guards.
    * **Schema Defaults**: Utilizes Drift null safety to prevent crashes on missing JSON keys.
    * **Error Tolerance**: Filters out invalid mood names while preserving valid data.

### `prompt_selection_system.dart`
A stochastic scoring algorithm used to rank prompts for the user.
* **Scoring Logic**: Prompts are weighted based on Novelty (+2.0), Category Preference (+2.0), Mood Alignment (+1.2 per match), and Difficulty Alignment ($2.0 - |prompt - user|$).
* **Selection**: Uses `_weightRank` (biased random function) to ensure variety and non-deterministic results.

### `adaptation_system.dart`
Learns and updates user profiles based on behavior.
* **Difficulty**: Adjusts based on completion rates (>80% to increment, <50% to decrement) and engagement metrics (word count/time).
* **Preferences**: Categories are marked "Preferred" if they reach a score $\ge 15$ based on participation, high engagement, and positive sentiment.
* **Avoidance**: Automatically adds Prompt IDs to an `avoidedPrompts` list if skipped $\ge 3$ times.

---

## 2. Curated Assets (`/assets`)

### `assets/prompts.json`
A library of **expert-created, clinically informed** journaling resources adapted for the app.
* **Sources**: Licensed clinicians and academic institutions including Charlie Health, Laura Copley (PhD, LMFT), Adonaca Barrett, and the University of Rochester Medical Center.
* **Therapeutic Frameworks**: Grounded in CBT, ACT, Mindfulness, and Positive Psychology.

### Data Schema
| Field | Description |
| :--- | :--- |
| `text` | The compassionate, action-oriented prompt. |
| `category` | Grouping used for preference learning (e.

Commit:
refactor: replace avoidedPrompts JSON with UserAvoidedPrompts join table

- Add UserAvoidedPrompts table with userId, promptId, and avoidedAt timestamp
- Remove avoidedPrompts text field from User table
- Add database migration from schema version 1 to 2
- Use proper relational design instead of JSON denormalization

Commit:
fix: update adaptation system to use UserAvoidedPrompts table

- Replace JSON-based avoidedPrompts with UserAvoidedPrompts table operations
- Add expiration mechanism: remove avoided prompts older than 30 days
- Optimize query: only fetch skipped interactions instead of all
- Fix preference learning: remove % 6 requirement, now runs when >= 6 entries
- Add error handling with try-catch in runAdaptation
- Add null check for user in _adjustDifficultyLevel

Commit:
fix: filter avoided prompts in prompt selection system

- Add _getAvoidedPromptIds method to query UserAvoidedPrompts table
- Filter out avoided prompts before scoring and ranking
- Add error handling with try-catch in rankPromptsForUser
- Ensure users don't see prompts they've skipped 3+ times

Commit:
fix: move prompts.json from assetes/ to assets/ and add to pubspec.yaml

- Move prompts.json from misspelled assetes/ directory to assets/
- Move README.md to assets/prompts_README.md
- Add assets/prompts.json to pubspec.yaml assets section
- Fixes asset loading issue in PromptInitializationService

Commit:
test: update tests for UserAvoidedPrompts schema change

- Update adaptation_system_test to query UserAvoidedPrompts table
- Remove avoidedPrompts field from database_test.dart
- Tests now verify join table relationship instead of JSON field

Commit:
chore: regenerate database code for UserAvoidedPrompts table

Commit:
fix: remove unsupported deleteColumn from migration

Drift doesn't support deleteColumn. The avoidedPrompts column
will be ignored since it's no longer in the schema.

Commit:
feat: auto-reset oldest avoided prompts when all are avoided

When all prompts are avoided, automatically reset the oldest 50%
of avoided prompts to ensure users always have prompts available.
This prevents the system from returning empty results.

Commit:
feat: add difficulty balancing to prevent consecutive same-difficulty prompts

Prevents more than 2 consecutive prompts of the same difficulty level.
Prioritizes prompts closer to user's current difficulty while ensuring variety.
This prevents overwhelming users with too many difficult prompts in a row.

Commit:
feat: integrate prompting system into GuidedMode

- Initialize prompts from JSON on app startup
- Create/get default user on app startup
- Replace hardcoded prompts with PromptSelectionSystem
- Track prompt interactions (completed/skipped) in database
- Run adaptation service after completing prompts
- Pass database and userId through widget tree
- Fix: Initialize prompts list immediately to prevent empty access errors
- Add loading state and better error handling with debug logging

This integrates the full prompting system with proper ranking,
filtering, difficulty balancing, and user preference learning.

Commit:
fix: update widget_test to work with new MyApp signature

- Create test database and user before initializing MyApp
- Remove outdated counter test
- Add proper cleanup
